hw/top_earlgrey/syn/chip_earlgrey_asic.sdc

```diff
@@ -182,82 +182,219 @@ if {!$IS_CDC_RUN} {
+# is sufficiently constrained in the 48 MHz case above.
+# The half-cycle sampling target frequency is 24 MHz. Over-constrain to 25 MHz.
+set SPI_TPM_TCK 40.0
+create_clock -name SPI_TPM_CLK -add -period ${SPI_TPM_TCK} [get_ports ${SPI_DEV_CLK_PIN}]
+set_clock_uncertainty ${SETUP_CLOCK_UNCERTAINTY} [get_clocks SPI_TPM_CLK]
+set_propagated_clock SPI_TPM_CLK
+
+create_generated_clock -name SPI_TPM_IN_CLK \
+    -source [get_ports ${SPI_DEV_CLK_PIN}] -divide_by 1 \
+    -master_clock SPI_TPM_CLK -add \
+    [get_pins top_earlgrey/u_spi_device/u_clk_spi_in_buf/clk_o]
+create_generated_clock -name SPI_TPM_OUT_CLK \
+    -source [get_ports ${SPI_DEV_CLK_PIN}] -divide_by 1 -invert \
+    -master_clock SPI_TPM_CLK -add \
+    [get_pins top_earlgrey/u_spi_device/u_clk_spi_out_buf/clk_o]
 
-set_output_delay ${SPI_DEV_OUT_DEL} [get_ports SPI_DEV_D0]   -clock SPI_DEV_CLK
-set_output_delay ${SPI_DEV_OUT_DEL} [get_ports SPI_DEV_D1]   -clock SPI_DEV_CLK
-set_output_delay ${SPI_DEV_OUT_DEL} [get_ports SPI_DEV_D2]   -clock SPI_DEV_CLK
-set_output_delay ${SPI_DEV_OUT_DEL} [get_ports SPI_DEV_D3]   -clock SPI_DEV_CLK
+# bidir ports
+set_input_delay -min ${SPI_DEV_IN_DEL_MIN} ${SPI_DEV_DATA_PORTS} \
+    -clock_fall -clock SPI_TPM_CLK -add_delay
+set_input_delay -max ${SPI_DEV_IN_DEL_MAX} ${SPI_DEV_DATA_PORTS} \
+    -clock_fall -clock SPI_TPM_CLK -add_delay
+
+# Half-cycle sampling has the host on the next rising edge.
+set_output_delay -min ${SPI_DEV_OUT_DEL_MIN} ${SPI_DEV_DATA_PORTS} \
+    -clock SPI_TPM_CLK -add_delay
+set_output_delay -max ${SPI_DEV_OUT_DEL_MAX} ${SPI_DEV_DATA_PORTS} \
+    -clock SPI_TPM_CLK -add_delay
+
+# SPI TPM CSB, the chip-select for TPM mode.
+# Any muxed port could be a SPI TPM CSB.
+set MUXED_IOA_PORTS [get_ports IOA*]
+set MUXED_IOB_PORTS [get_ports IOB*]
+set MUXED_IOC_PORTS [get_ports IOC*]
+set MUXED_IOR_PORTS [get_ports "IOR0 IOR1 IOR2 IOR3 IOR4 IOR5 IOR6 IOR7 IOR10 IOR11 IOR12 IOR13"]
+set ALL_MUXED_PORTS "${MUXED_IOA_PORTS} ${MUXED_IOB_PORTS} ${MUXED_IOC_PORTS} ${MUXED_IOR_PORTS}"
+
+# TPM CSB input delays.
+set_input_delay -min ${SPI_DEV_IN_DEL_MIN} [get_ports ${ALL_MUXED_PORTS}] \
+    -clock SPI_TPM_CLK -clock_fall -add_delay
+set_input_delay -max ${SPI_DEV_IN_DEL_MAX} [get_ports ${ALL_MUXED_PORTS}] \
+    -clock SPI_TPM_CLK -clock_fall -add_delay
+
+# Relax hold path for TPM CSB, since CSB changes nominally on the same edge as
+# SPI_TPM_OUT_CLK, but the latter isn't actually toggling.
+set_multicycle_path -hold -end -from [get_clocks SPI_TPM_CLK] \
+    -through [get_ports ${ALL_MUXED_PORTS}] \
+    -to [get_clocks SPI_TPM_OUT_CLK] 1
+
+# Remove hold analysis from the following paths to ports. Even though the pins
+# can change before the prior data was latched upstream, their effect is held
+# back by other logic on SPI_TPM_OUT_CLK.
+# Note: The final output logic equation must not permit glitches in the presence
+# of changes on the listed pins. Otherwise, any hold time failures could be
+# real.
```

Comment:
Is there a chance that the synthesis tool may get "smart" and rearrange this logic so that glitches can propagate? Or is it something that is straightforward (e.g. an AND gate and no reconvergent paths)?

Reply:
I'm not sure. If it ends up as written, there would be a final mux that only selects a path from SPI_TPM_IN_CLK for the first bit of a byte, and the mux selection controls only change on SPI_TPM_OUT_CLK.

However, there are a lot of signals that make up the path on that mux's SPI_TPM_IN_CLK input side...

Reply:
Ah this is basically this issue here: https://github.com/lowRISC/opentitan/issues/18687

I see that the data mux select is computed from a counter. 
We can look at a synthesized netlist to double check whether this synthesized as written.